pennylane/workflow/qnode.py

```diff
@@ -1051,29 +1051,35 @@ def _execution_component(self, args: tuple, kwargs: dict, override_shots) -> qml
                 "Cannot use the 'one-shot' method for mid-circuit measurements with analytic mode."
             )
 
-        # Add the device program to the QNode program
+        full_transform_program = qml.transforms.core.TransformProgram(self.transform_program)
+        inner_transform_program = qml.transforms.core.TransformProgram()
+        config = None
+
         if isinstance(self.device, qml.devices.Device):
+
             config = _make_execution_config(self, self.gradient_fn)
             device_transform_program, config = self.device.preprocess(execution_config=config)
-            full_transform_program = self.transform_program + device_transform_program
-        else:
-            config = None
-            full_transform_program = qml.transforms.core.TransformProgram(self.transform_program)
+
+            if config.use_device_gradient:
+                full_transform_program += device_transform_program
+            else:
+                inner_transform_program += device_transform_program
 
         has_mcm_support = (
             any(isinstance(op, MidMeasureMP) for op in self._tape)
             and hasattr(self.device, "capabilities")
             and self.device.capabilities().get("supports_mid_measure", False)
         )
         if has_mcm_support:
-            full_transform_program.add_transform(
+            inner_transform_program.add_transform(
                 qml.devices.preprocess.mid_circuit_measurements,
                 device=self.device,
                 mcm_config=self.execute_kwargs["mcm_config"],
+                interface=self.interface,

```

Comment:
This should cover `lightning.kokkos`'s mid-circuit measurement pipeline 😄

Reply:
Is `lightning.kokkos` a legacy device? If so then yes I guess.